Documentation for Taipy Studio Configuration Builder (#319) (#364)
Avaiga/taipy-doc: docs/manuals/studio/config.md

# Editing configuration files

Applications that rely on [Taipy Core](../core/index.md) need to define
a [Configuration](../core/config/index.md) where the data flow is
described.<br/>

> This configuration can be coded in Python, tuned using environment variables,
> or be expressed as a TOML file as described in the
> [TOML file configuration](../core/config/advanced-config.md#explicit-toml-file-configuration)
> section.

Reviewer: Two remarks: 

1. Vincent and I recently made some updates to the page you are referring to. I just pushed the changes. You link is no longer up-to-date.
 
2. What you are saying is not really the truth. The section you mention is made to document how to override a python code configuration using TOML file, And not how to build a complete configuration in TOML.

I would not say that we can build a configuration using TOML. I would just say either we use Python code, or we use Studio and load the TOML result generated by Studio.

```suggestion
This configuration can be coded in Python, tuned using environment variables
(as described in the [advanced configuration](../core/config/advanced-config.md#Python-code-configuration) section).
```

> Taipy Studio provides ways to design the Taipy Configuration files for your applications.

Reviewer: ```suggestion
Taipy Studio provides graphical ways to design the Taipy Configuration files for your applications.
```

Within the Visual Studio Code environment, you can create configuration elements (for
Data Nodes, Tasks, Pipelines, and Scenarios), configure those elements, and use a visual
representation of those configuration elements as a graph.

<p align="center">
  <img src="../images/config_overview.png" width=75%>
</p>

## Introduction

Taipy Studio has a dedicated panel called "Taipy Configs" where mostly everything
happens. To open this panel, use the "View > Open View..." menu option and search
for the view called "Taipy Configs".

The view appears like this:

<p align="center">
  <img src="../images/config_panel.png" width=30%>
</p>

This panel is split into several sections that let the user access all the
configuration elements:

- Config Files: holds a list of configuration files (`*.toml`) that exist in the
    project.<br/>
    The user can select a configuration file in this list to populate the other
    sections of the Taipy Configs view.
- Data Notes: holds the list of
    [Data Node configurations](../core/config/data-node-config.md) defined

>     in the selected configuration files.<br/>

Reviewer 2: files => file

Author: 👌

    The user can select a Data Node configuration in this list, to access all its
    properties in the Details section (see below).
- Tasks: holds the list of
    [Task configurations](../core/config/task-config.md) defined

>     in the selected configuration files.<br/>

Reviewer 2: idem

Author: idem

    [Pipeline configurations](../core/config/pipeline-config.md) defined
    in the selected configuration files.<br/>
    The user can select a Pipeline configuration in this list to access all its
    properties in the Details section (see below).
- Scenarios: holds the list of 
    [Scenario configurations](../core/config/scenario-config.md) defined
    in the selected configuration files.<br/>
    The user can select a Scenario configuration in this list to access all its
    properties in the Details section (see below).
- Details: When a configuration element is selected, the Details section displays
    all the properties of this configuration element and allows the user to
    modify these properties.

Using these different sections, you can add, rename or remove configuration
elements, as well as change their properties in the "Details" section.

All your actions impact the selected configuration file;
you can use Visual Studio Code's undo/redo feature if needed.

## Managing configuration files

The "Config Files" section holds the list of configuration files (`*.toml`) in your
project. You will select the one you want to edit from that list.<br/>

> This list shows the root names of all the configuration files. If two configuration files

Reviewer 2: root name => base name ?

Author: Yes! That's far better thanks.

have the same name (but are located in two different directories), the path to the directory
where these files belong will be displayed next to the root file name to avoid
confusion.

Because the "Config Files" list is synchronized with the files included in your project,
adding a configuration file is simply a matter of creating it from the Visual Studio
Code Explorer panel:

<p align="center">
  <img src="../images/config_init.gif" width=80%>
</p>

Similarly, if you remove or rename a configuration file from the Explorer panel of Visual
Studio Code, the change is immediately reflected in the Config Files section of the
Taipy Configs panel.

Note that if you right-click a configuration file, Taipy Studio displays a menu that
let you choose one of two options:

- "Reveal file in explorer view": Visual Studio Code will select and focus on the file
    item in the project's files in the Explorer pane.

> - "Show view": creates a view on the configuration file that represents the entire content

Reviewer 2: on the configuration file => from the configuration file ?
Add the location of the created view ? editor center space ...

configuration file.

## The graph view

## Text edition of the configuration file

A configuration file is a standard text file, with a '`.toml`' extension. Visual Studio
Code lets you edit it manually, as any other text file.<br/>
However, Taipy Studio provides some support for identifying problems in configuration
files and accelerating typing thanks to auto-completion features.

### Auto-completion

If you are entering text in a configuration section ([DATA_NODE.*id*], [TASK.*id*], etc.),
you can press the `Ctrl-<SPACE>` key combination, then Taipy Studio will suggest relevant
fragments that would prevent you from typing them entirely.

Those suggestions include property names and values, as demonstrated in the following animation,
showing how we could create a new Data Node configuration in the body of a configuration file:

<p align="center">
  <img src="../images/config_autocomplete.gif" width=80%>
</p>

You will also get support for auto-completion in other contexts, such as a reference
to another configuration element.

### Spotting problems

Taipy Studio helps point out potential problems in your configuration files that are

> indicated both in the body of the file content and in the Problems pane.

Reviewer 2: and in the config details view

Author: Ok!
